### PR TITLE
PoC user feed migration

### PR DESCRIPTION
**Pull request by nobuti**
This PR migrates the user feed view in order to decouple cartodb.js from the core.

# Files migration

## Config

cdb.config cartodb/config.js - dashboard/data/config-model.js

## Url helpers

cdb.common.Url cartodb/old_common/urls/url.js - dashboard/data/url-model.js
cdb.common.UserUrl cartodb/old_common/urls/user_url.js - dashboard/data/user-url-model.js
cdb.common.DashboardUrl cartodb/old_common/urls/dashboard_url.js - dashboard/data/dashboard-url-model.js
cdb.common.DashboardDatasetsUrl cartodb/old_common/urls/dashboard_datsets_url.js - dashboard/data/dashboard-datsets-url-model.js
cdb.common.DashboardVisUrl cartodb/old_common/urls/dashboard_vis_url.js - dashboard/data/dashboard-vis-url-model.js

## Dropdown view and modifications

cdb.ui.common.Dropdown [cartodb.js] src/ui/common/dropdown.js - dashboard/common/dropdown-base-view.js
cdb.admin.DropdownMenu cartodb/old_common/dropdown_menu - dashboard/common/dropdown-admin-view.js

## Backbone sync options

Backbone sync options from cartodb/app.js - dashboard/data/backbone/sync-options.js

cdb.open.AuthenticatedUser cartodb/public/authenticated_user.js - dashboard/data/authenticated-user-model

## Like

cdb.admin.Like cartodb/models/like.js - dashboard/data/like-model.js
cartodb/common/views/likes/view.js - dashboard/common/likes/like-view.js
cartodb/common/views/likes/template.jst.ejs - dashboard/common/likes/like.tpl

## View factory

Missing dropdowns factory views, it's likely that it will need to be patched for some entry points:
cartodb/common/view_factory.js - cartodb3/components/view-factory.js

## Navigation menu dropdowns

cartodb/public_common/user_industries/dropdown_view.js - dashboard/public-common/user-industries/dropdown-view.js
cartodb/public_common/user_industries/dropdown_template.jst.ejs - dashboard/public-common/user-industries/dropdown.tpl
cartodb/public_common/user_settings/dropdown_view.js - dashboard/public-common/user-settings/dropdown-view.js
cartodb/public_common/user_settings/dropdown_template.jst.ejs - dashboard/public-common/user-settings/dropdown.tpl
cartodb/public_common/user_industries_view.js - dashboard/public-common/user-industries-view.js
cartodb/public_common/user_settings_view.js - dashboard/public-common/user-settings-view.js
cartodb/public_common/user_settings_template.jst.ejs - dashboard/public-common/user-settings.tpl

## User feed inner views

cartodb/common/create_default_fallback_map.js - dashboard/common/create-fallback-map.js
cartodb/common/views/mapcard_preview.js - dashboard/common/mapcard-preview-view.js
cartodb/public_dashboard/fav_map_view.js - dashboard/public-dashboard/fav-map-view.js

## User feed view and templates (aka the timeline/tree)

cartodb/user_feed/view.js - dashboard/public-dashboard/feed-view.js
cartodb/user_feed/dataset_item_template.jst.ejs - dashboard/public-dashboard/user-feed/dataset-item.tpl
cartodb/user_feed/empty.jst.ejs - dashboard/public-dashboard/user-feed/empty.tpl
cartodb/user_feed/feed_collection.js - dashboard/public-dashboard/user-feed/feed-collection.js
cartodb/user_feed/template.jst.ejs - dashboard/public-dashboard/user-feed/feed.tpl
cartodb/user_feed/loading.jst.ejs - dashboard/public-dashboard/user-feed/loading.tpl
cartodb/user_feed/map_item_template.jst.ejs - dashboard/public-dashboard/user-feed/map-item.tpl

## Different helpers

cartodb/common/view_helpers/random_quote.js - cartodb3/components/loading/random-quote.js
cartodb/common/views/scrollable_header - dashboard/util/scroll-tofixed-view

## User feed entry point

cartodb/user_feed/entry.js - dashboard/user-feed.js

## Special attention required for the user model.

dashboard/data/user-model is extending cartodb3/models/user-model. The old model includes layers relation that I'm almost 100% sure it's not needed for the dashboard. As today, it's working fine for the user-feed entry point, but in the future, we need to pay attention to add/overwrite/change some methods or properties needed for other entry points.

# Build

Inside webpack folder, webpack config files are namespaced inside folder for builder, dashboard and carto-node. As this being a PoC, the next steps will make the full integration of the deployment system once validated.

# Development

`grunt dashboard` mimics `grunt dev`
`grunt dashboard_specs` run the tests ~~in the console~~.

# Workflow

There are some points you need to understand before getting your hands dirty:

- In Builder, we avoid globals in favor of dependency injection. So we should follow the same approach here. You need to pass down the components' tree every dependency required for a component.
- There is no cdb global namespace, so no more god or global bus for event management. Instead, a global bus, listen to views, models and collections events.
- We are inheriting from Backbone.Model directly, there is not a custom model anymore. This is important because in the old code the core model had some custom events (see https://github.com/CartoDB/cartodb.js/blob/1e770ede4d2764181104322d4d7e5ba9ce393455/src/core/model.js#L37) you need to move to standard events.
- At model level, prioritize model already migrated. Always check the data structure integrity and value the cost of adapting/inheriting. This is important for core models as User model, Config model and so.

So taking all of this into account, I started on the entry point. I basically made a list of the dependencies every dependency had, local and global (cdb.wadus). For every one of them, I did the same over and over again until I had got the full dependency tree covered. In order to migrate the files:

- views inherit from backbone/core-view
- models and collections inherit from Backbone.Model and Backbone.Collection
- When you found cdb.wadus, use the list above of migrated files, it's likely you will need it.

As part of this PoC, I made only a couple tests, I focused on migrating the existing tests. But the approach was the same: inject every dependency required to not rely on globals.

# Bugs

When created the favorite map, we are disabling the scroll wheel option, overwriting the vizjson's option. But the thing is, this is not working in cartodb.js v4 right now: https://github.com/CartoDB/cartodb.js/issues/2007

# One more thing

Raster is forced in the favorite map in order to not add tangram script. Tangram will be deprecated shortly so it doesn't make sense to include.

### CODE REVIEW

**Comment by elenatorro** [comment]
VAMOS 💥 ✨

**Comment by ivanmalagon** [comment]
Before passing to acceptance we need to do two things:
- Serve old or new assets to the user depending on the flag `builder_enabled`.
- Fix the scroll wheel bug.

**Comment by rubenmoya** [comment]
We've also found two more things:
- When the user has no maps, the user_feed header breaks
- When the user has no maps and the default basemap is google maps, the script is not loaded

File: app/controllers/admin/pages_controller.rb [hunk 1]
```diff
@@ -129,6 +129,7 @@ def user_feed
       @maps_count         = maps_builder.build.count
       @website            = website_url(@viewed_user.website)
       @website_clean      = @website ? @website.gsub(/https?:\/\//, "") : ""
+      @hasNewDashboard    = @viewed_user.builder_enabled && viewed_user.has_feature_flag?('dashboard_migration')
```
[the remark this answers is not in the document]
Comment (rjimenezda) [reply app/controllers/admin/pages_controller.rb]: mimimimi

**Comment by rjimenezda** [comment]
We changed the name of the new assets, because they were overwriting the old user_feed.js & builder's common.js

We also added the feature flag 'dashboard_migration', so only if a user has that enabled and builder will see the new user feed.

On the migrated version, the 'fav map' for the case where the user has no maps is handled entirely on JS now.

The issue with the scroll & the map is pending CR but should be released :soon: 

You can see it on ded14, ruben-ded14 has the flag. It won't work because the new assets are not being generated on prod builds, but you can see how the new assets are being requested.

Things missing: 

- [ ] Building the new assets during a regular build
- [ ] Do whatever is required to have the user flag available on prod

**Pull request merged**